resolver_test.go

```diff
@@ -165,3 +167,104 @@ func TestDNSIPQuery(t *testing.T) {
 	w.ClearResponse()
 
 }
+
+var requests = 0
+
+func handleTestDNSRequestServFailOnce(w dns.ResponseWriter, r *dns.Msg) {
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Compress = false
+	if requests == 0 {
+		m.SetRcode(r, dns.RcodeServerFailure)
+	}
+	requests = requests + 1
+	w.WriteMsg(m)
```

Comment:
The global state here and the comparison against it seems a little odd.  The way TestDNSProxyServFail below compared "requests" w/ the previous value threw me at first.  Take it for what it's worth, but suggest: 
```go
func newHandleTestDNSRequestServFailOnce(requests *int) func(dns.ResponseWriter, dns.*Msg) {
	return func(w dns.ResponseWriter, r *dns.Msg) {
		...
		*requests = *requests + 1
		...
	}
}
...
func TestDNSProxyServFail(t *testing.T) {
...
   var nRequests int = 0
   dns.HandleFunc(".", newHandleTestDNSRequestServFailOnce(&nRequests))
...
   if nRequests != 2 {
...
   }
...
}
```
Just thought it might help readability.  (One could also embed the whole newHandleTestDNSRequestServFailOnce() body in the main test function and not have to pass the request counter pointer....  but that seems uglier, really.)

Reply:
Thanks for the suggestion! I did not like that global either and your suggestion is a decent way to keep it bound to the scope of the function. Will incorporate it.

Reply:
+1 to this